scipy/sparse/linalg/eigen/arpack/tests/test_arpack.py

```diff
@@ -790,11 +795,13 @@ def reorder(args):
         A = np.random.RandomState(1909).randn(n, m)
         L = CheckingLinearOperator(A)
 
-        for solver in [None, 'arpack', 'lobpcg']:
+        for solver in [None, 'arpack', 'lobpcg', 'propack']:
             U1, s1, VH1 = reorder(svds(A, k, which="SM", solver=solver))
             U2, s2, VH2 = reorder(svds(L, k, which="SM", solver=solver))
 
-            assert_allclose(np.abs(U1), np.abs(U2))
+            if solver != 'propack':
+                # PROPACK's U is variable run-to-run
```

Comment:
Running PROPACK on the same array multiple times using the same random seed produces variable `U`, so I assume this is a "feature" of PROPACK and not a bug

Reply:
Unless there is an option we can set, I guess we have to live with that. Maybe in that case we shouldn't expose on option that claims to control the seed, and we should document that the output is not deterministic and cannot be controlled with random seeds.

Reply:
I see what's going on here, the test is incorrect.  When `which="SM"` is tested, `v0` is not set.  This is different than on lines 785-6 where `v0` is provided to avoid the random seeding behavior.  When `v0` is supplied this test succeeds as expected

Reply:
"Fixing" this test also breaks arpack, so that's a problem:
```
scipy/sparse/linalg/eigen/arpack/tests/test_arpack.py .....................F.........                                                                                                               [100%]

================================================================================================ FAILURES =================================================================================================
_____________________________________________________________________________________________ test_svd_linop ______________________________________________________________________________________________
ValueError: 0-th dimension must be fixed to 27 but got 15


The above exception was the direct cause of the following exception:

    def test_svd_linop():
        nmks = [(6, 7, 3),
                (9, 5, 4),
                (10, 8, 5)]
    
        def reorder(args):
            U, s, VH = args
            j = np.argsort(s)
            return U[:, j], s[j], VH[j, :]
    
        for n, m, k in nmks:
            # Test svds on a LinearOperator.
            A = np.random.RandomState(52).randn(n, m)
            L = CheckingLinearOperator(A)
    
            for solver in [None, 'arpack', 'lobpcg', 'propack']:
                if solver == 'propack':
                    v0 = np.ones(n)
                else:
                    v0 = np.ones(min(A.shape))
    
                U1, s1, VH1 = reorder(svds(A, k, v0=v0, solver=solver))
                U2, s2, VH2 = reorder(svds(L, k, v0=v0, solver=solver))
    
                assert_allclose(np.abs(U1), np.abs(U2))
                assert_allclose(s1, s2)
                assert_allclose(np.abs(VH1), np.abs(VH2))
                assert_allclose(np.dot(U1, np.dot(np.diag(s1), VH1)),
                                np.dot(U2, np.dot(np.diag(s2), VH2)))
    
            # Try again with which="SM".
            A = np.random.RandomState(1909).randn(n, m)
            L = CheckingLinearOperator(A)
    
            for solver in [None, 'arpack', 'lobpcg', 'propack']:
>               U1, s1, VH1 = reorder(svds(A, k, v0=v0, which="SM", solver=solver))

A          = array([[-0.61687924,  0.25336626,  0.68918087,  2.94626411,  0.08939153],
       [-1.84290537,  1.34520453,  1.6625265...292, -0.73941863, -0.26401395,  1.16853114],
       [-2.69822328,  0.10640379,  1.06363414, -0.02626475,  0.37829858]])
L          = <9x5 CheckingLinearOperator with dtype=float64>
U1         = array([[-0.62902399,  0.08437264,  0.19657502, -0.2439448 ],
       [-0.03938899,  0.28737564,  0.53104249,  0.3203668...    [ 0.35546364,  0.09106982, -0.26336457,  0.2236074 ],
       [-0.26109191, -0.34163455, -0.13515346,  0.47641654]])
U2         = array([[-0.62902399,  0.08437264,  0.19657502, -0.2439448 ],
       [-0.03938899,  0.28737564,  0.53104249,  0.3203668...    [ 0.35546364,  0.09106982, -0.26336457,  0.2236074 ],
       [-0.26109191, -0.34163455, -0.13515346,  0.47641654]])
VH1        = array([[-0.81194723,  0.4760018 , -0.28137609, -0.11372536, -0.14851942],
       [ 0.28584573,  0.44076879,  0.3982669...88 ,  0.32070221, -0.50716985,  0.75891623],
       [ 0.20810083,  0.72255691,  0.22153587,  0.48636748,  0.38597162]])
VH2        = array([[-0.81194723,  0.4760018 , -0.28137609, -0.11372536, -0.14851942],
       [ 0.28584573,  0.44076879,  0.3982669...88 ,  0.32070221, -0.50716985,  0.75891623],
       [ 0.20810083,  0.72255691,  0.22153587,  0.48636748,  0.38597162]])
dt         = <class 'numpy.complex64'>
eps        = 0.001
k          = 4
m          = 5
n          = 9
nmks       = [(6, 7, 3), (9, 5, 4), (10, 8, 5)]
reorder    = <function test_svd_linop.<locals>.reorder at 0x7ff645fc04c0>
rng        = RandomState(MT19937) at 0x7FF64603B140
s1         = array([1.59693189, 2.02249862, 2.52560135, 4.67197394])
s2         = array([1.59693189, 2.02249862, 2.52560135, 4.67197394])
solver     = None
v0         = array([1., 1., 1., 1., 1., 1., 1., 1., 1.])

scipy/sparse/linalg/eigen/arpack/tests/test_arpack.py:799: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
scipy/sparse/linalg/eigen/arpack/arpack.py:1878: in svds
    eigvals, eigvec = eigsh(XH_X, k=k, tol=tol ** 2, maxiter=maxiter,
        A          = array([[-0.61687924,  0.25336626,  0.68918087,  2.94626411,  0.08939153],
       [-1.84290537,  1.34520453,  1.6625265...292, -0.73941863, -0.26401395,  1.16853114],
       [-2.69822328,  0.10640379,  1.06363414, -0.02626475,  0.37829858]])
        XH_X       = <5x5 _CustomLinearOperator with dtype=float64>
        XH_dot     = <built-in method dot of numpy.ndarray object at 0x7ff646010ab0>
        XH_mat     = <built-in method dot of numpy.ndarray object at 0x7ff646010ab0>
        X_dot      = <built-in method dot of numpy.ndarray object at 0x7ff646010d50>
        X_matmat   = <built-in method dot of numpy.ndarray object at 0x7ff646010d50>
        k          = 4
        largest    = False
        m          = 5
        matmat_XH_X = <function svds.<locals>.matmat_XH_X at 0x7ff645fc01f0>
        matvec_XH_X = <function svds.<locals>.matvec_XH_X at 0x7ff645fc05e0>
        maxiter    = None
        n          = 9
        ncv        = None
        return_singular_vectors = True
        solver     = None
        tol        = 0
        v0         = array([1., 1., 1., 1., 1., 1., 1., 1., 1.])
        which      = 'SM'
scipy/sparse/linalg/eigen/arpack/arpack.py:1690: in eigsh
    params.iterate()
        A          = <5x5 _CustomLinearOperator with dtype=float64>
        M          = None
        M_matvec   = None
        Minv       = None
        Minv_matvec = None
        OPinv      = None
        k          = 4
        matvec     = <bound method LinearOperator.matvec of <5x5 _CustomLinearOperator with dtype=float64>>
        maxiter    = None
        mode       = 1
        n          = 5
        ncv        = None
        params     = <scipy.sparse.linalg.eigen.arpack.arpack._SymmetricArpackParams object at 0x7ff64603d700>
        return_eigenvectors = True
        sigma      = None
        tol        = 0
        v0         = array([1., 1., 1., 1., 1., 1., 1., 1., 1.])
        which      = 'SM'
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <scipy.sparse.linalg.eigen.arpack.arpack._SymmetricArpackParams object at 0x7ff64603d700>

    def iterate(self):
        self.ido, self.tol, self.resid, self.v, self.iparam, self.ipntr, self.info = \
>           self._arpack_solver(self.ido, self.bmat, self.which, self.k,
                                self.tol, self.resid, self.v, self.iparam,
                                self.ipntr, self.workd, self.workl, self.info)
E       ValueError: failed in converting 10th argument `workd' of _arpack.dsaupd to C/Fortran array

self       = <scipy.sparse.linalg.eigen.arpack.arpack._SymmetricArpackParams object at 0x7ff64603d700>

scipy/sparse/linalg/eigen/arpack/arpack.py:536: ValueError
========================================================================================= short test summary info =========================================================================================
FAILED scipy/sparse/linalg/eigen/arpack/tests/test_arpack.py::test_svd_linop - ValueError: failed in converting 10th argument `workd' of _arpack.dsaupd to C/Fortran array
====================================================================================== 1 failed, 30 passed in 1.23s =======================================================================================
 UNVARIANT SUBSPACE, beta small
 UNVARIANT SUBSPACE, beta small

```